pmd-java/src/main/resources/category/java/codestyle.xml

```diff
@@ -1351,6 +1351,40 @@ public class ClassInDefaultPackage {
         </example>
     </rule>
 
+    <rule name="NumericLiteralConvention"
+          language="java"
+          since="6.9.0"
+          minimumLanguageVersion="1.7"
+          message="Numeric literals with more than 3 digits must use '_' as a separator"
+          class="net.sourceforge.pmd.lang.rule.XPathRule"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_codestyle.html#numericliteralconvention">
+        <description>
+            Numeric literals with more than 3 digits must use '_' as a separator.
+        </description>
+        <priority>3</priority>
+        <properties>
+            <property name="version" value="2.0"/>
+            <property name="xpath">
+                <value>
+                    <![CDATA[
+//Literal[@IntLiteral = true() or
+ @LongLiteral = true() or
+ @DoubleLiteral = true() or
+ @FloatLiteral = true()]
+ [not(matches(@Image, "^[0-9]{1,3}(_[0-9]{3})*(l|L|\.[0-9]+)?(d|D|f|F)?$"))]
```

Comment:
One alternation may be simplified:

```suggestion
 [not(matches(@Image, "^[0-9]{1,3}(_[0-9]{3})*(l|L|\.[0-9]+)?[dDfF]?$"))]
```

There should probably be a configurable minimum length threshold before the underscores are required. For instance for 4 digit numbers: I don't see a problem reading `1852`, whereas `1_852` is slightly less readable IMO.

Reply:
Could you let me know how configuring the length can be achieved?

Reply:
@rajeshggwp Yes, using a property is the way to go. We could split the string before and after the decimal separator, removing the `dfl` suffixes. Then each segment is *valid* iff either 
* it is shorter than the acceptable length, and either doesn't have any underscores, or all its underscores are well placed
* it is longer than the acceptable length, and all its underscores are well placed

If either segment is invalid, then we report the node. The following condition does the trick:
```ruby
[
  some $num in tokenize(@Image, "[.dDfFlL]") 
  satisfies not(
                 string-length($num) <= $acceptableLength 
                   and (
                         not(contains($num,"_")) 
                         or matches($num, "^[0-9]{1,3}(_[0-9]{3})*$")
                       )
		 or matches($num, "^[0-9]{1,3}(_[0-9]{3})*$")
               )
 ]
```